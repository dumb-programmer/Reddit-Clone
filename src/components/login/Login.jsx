import { useEffect, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import AppleButton from "./AppleButton";
import GoogleButton from "./GoogleButton";
import LoadingSVG from "../LoadingSVG";
import LoginAndSignUpLayout from "../LoginAndSignUpLayout";
import { loginUsingUsernameAndPassword } from "../../firebase";
import AlreadyLoggedInMessage from "./AlreadyLoggedInMessage";
import useAuthContext from "../../hooks/useAuthContext";

const Login = () => {
  const [data, setData] = useState({
    username: "",
    password: "",
  });

  const [loading, setLoading] = useState(false);
  const [errors, setErrors] = useState({
    username: false,
    password: false,
  });
  const navigate = useNavigate();
  const authenticated = useAuthContext();

  const handleInput = (e) => {
    setData({
      ...data,
      [e.target.name]: e.target.value,
    });
    setErrors({
      [e.target.name]: false,
    });
  };

  const handleFormSubmit = async (e) => {
    e.preventDefault();
    setLoading(true);
    const val = await loginUsingUsernameAndPassword(data);
    if (val === -1) {
      setErrors({
        ...errors,
        username: true,
      });
    } else if (val === -2) {
      setErrors({
        ...errors,
        password: true,
      });
    } else {
      navigate("/");
    }
    setLoading(false);
  };

  const handleUsernameBlur = (e) => {
    setErrors({
      ...errors,
      username: false,
    });
  };

  const handlePasswordBlur = (e) => {
    setErrors({
      password: false,
    });
  };

  useEffect(() => {
    if (authenticated) {
      setTimeout(() => navigate("/"), 1000);
    }
  }, [authenticated, navigate]);

  const disableLoginBtn = Object.values(errors).some((error) => error);

  if (authenticated) {
    return (
      <LoginAndSignUpLayout>
        <AlreadyLoggedInMessage />
      </LoginAndSignUpLayout>
    );
  }

  return (
    <LoginAndSignUpLayout>
      <h4>Log in</h4>
      <p className="user-agreement smaller">
        By continuing, you agree to our User Agreement and Privacy Policy.
      </p>
      <form
        className="login-form"
        onSubmit={
          !(loading || disableLoginBtn)
            ? handleFormSubmit
            : (e) => {
                e.preventDefault();
              }
        }
      >
        <GoogleButton />
        <AppleButton />
        <input
          className={`form-input ${errors.username ? "form-input__error" : ""}`}
          type="text"
          name="username"
          placeholder="Username"
          value={data.username}
          onChange={handleInput}
          onBlur={handleUsernameBlur}
          autoComplete="off"
          required
        />
        {errors.username && (
          <div className="error-message">Incorrect username</div>
        )}
        <input
          className={`form-input ${errors.password ? "form-input__error" : ""}`}
          type="password"
          name="password"
          placeholder="Password"
          value={data.password}
          onChange={handleInput}
          onBlur={handlePasswordBlur}
          required
        />
        {errors.password && (
          <div className="error-message">Incorrect password</div>
        )}
        <button
          type="submit"
          data-testid="login-btn"
          className={`submit-btn ${loading ? "btn__loading" : ""}`}
          disabled={errors.emailAlreadyRegistered}
        >
          {!loading ? "Log In" : <LoadingSVG height={35} width={35} />}
        </button>
        <p className="smaller">Forgot your username or password ?</p>
        <p className="smaller">
          New to Reddit? <Link to="/signup">Sign up</Link>
        </p>
      </form>
    </LoginAndSignUpLayout>
  );
};

export default Login;
