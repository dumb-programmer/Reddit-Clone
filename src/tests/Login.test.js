import { MemoryRouter } from "react-router-dom";
import Login from "../components/Login";
import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import * as Firebase from "../firebase";

jest.mock("../firebase.js");

Firebase.loginUsingUsernameAndPassword = jest.fn(async ({ username, password }) => {
    if (username === "elliot") {
        if (password === "123") {
            return 0;
        }
        return -2;
    }
    return -1;
});

describe("Login works", () => {
    test("Render", async () => {
        // TODO: custom render function
        const { container } = render(<MemoryRouter>
            <Login />
        </MemoryRouter>)

        expect(container).toMatchSnapshot();
    })

    test("Username input works", async () => {
        render(<MemoryRouter>
            <Login />
        </MemoryRouter>);
        const user = userEvent.setup();

        await user.type(screen.getByPlaceholderText("Username"), "testing");
        expect(screen.getByPlaceholderText("Username")).toHaveValue("testing");
    });

    test("Password input works", async () => {
        render(<MemoryRouter>
            <Login />
        </MemoryRouter>);
        const user = userEvent.setup();

        await user.type(screen.getByPlaceholderText("Password"), "testing");
        expect(screen.getByPlaceholderText("Password")).toHaveValue("testing");
    });

    test("Username and password are required", async () => {
        render(<MemoryRouter>
            <Login />
        </MemoryRouter>);
        const user = userEvent.setup();

        await user.click(screen.getByTestId("login-btn"));
        expect(screen.getByPlaceholderText("Username")).toBeInvalid();
        expect(screen.getByPlaceholderText("Password")).toBeInvalid();
    });

    test("Login button works", async () => {
        render(<MemoryRouter>
            <Login />
        </MemoryRouter>);
        const user = userEvent.setup();

        await user.type(screen.getByPlaceholderText("Username"), "test");
        await user.type(screen.getByPlaceholderText("Password"), "1234");
        await user.click(screen.getByTestId("login-btn"));

        expect(Firebase.loginUsingUsernameAndPassword).toHaveBeenCalled();
        expect(Firebase.loginUsingUsernameAndPassword).toHaveBeenCalledWith({ username: "test", password: "1234" });
    });

    test("Error is displayed for incorrect username", async () => {
        render(<MemoryRouter>
            <Login />
        </MemoryRouter>);
        const user = userEvent.setup();

        const username = "test";
        const password = "123";
        await user.type(screen.getByPlaceholderText("Username"), username);
        await user.type(screen.getByPlaceholderText("Password"), password);
        await user.click(screen.getByTestId("login-btn"));

        expect(await Firebase.loginUsingUsernameAndPassword(username, password)).toBe(-1);
        expect(screen.getByPlaceholderText(/username/i)).toHaveClass("form-input__error");
        expect(screen.getByText(/incorrect username/i)).toBeInTheDocument();
        expect(screen.getByText(/incorrect username/i)).toHaveClass("error-message");
    });

    test("Error is displayed for incorrect password", async () => {
        render(<MemoryRouter>
            <Login />
        </MemoryRouter>);
        const user = userEvent.setup();

        const username = "elliot";
        const password = "34";
        await user.type(screen.getByPlaceholderText("Username"), username);
        await user.type(screen.getByPlaceholderText("Password"), password);
        await user.click(screen.getByTestId("login-btn"));

        expect(await Firebase.loginUsingUsernameAndPassword({ username, password })).toBe(-2);
        expect(screen.getByPlaceholderText(/password/i)).toHaveClass("form-input__error");
        expect(screen.getByText(/incorrect password/i)).toBeInTheDocument();
        expect(screen.getByText(/incorrect password/i)).toHaveClass("error-message");
    });

    test("User is logged in on correct username and password", async () => {
        render(<MemoryRouter>
            <Login />
        </MemoryRouter>);
        const user = userEvent.setup();

        const username = "elliot";
        const password = "123";
        await user.type(screen.getByPlaceholderText("Username"), username);
        await user.type(screen.getByPlaceholderText("Password"), password);
        await user.click(screen.getByTestId("login-btn"));

        expect(await Firebase.loginUsingUsernameAndPassword({ username, password })).toBe(0);
    });
})