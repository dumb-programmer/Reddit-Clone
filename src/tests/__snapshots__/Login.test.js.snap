// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login works Render 1`] = `
<div>
  <div
    class="login-page"
  >
    <aside
      class="artwork"
    />
    <main
      class="login-and-signup"
    >
      <h4>
        Log in
      </h4>
      <p
        class="user-agreement smaller"
      >
        By continuing, you agree to our User Agreement and Privacy Policy.
      </p>
      <form
        class="login-form"
      >
        <div
          class="login-provider-btn-container"
        >
          <span
            class="provider-logo"
            id="google-logo"
          >
            <svg
              height="18px"
              preserveAspectRatio="xMidYMid"
              version="1.1"
              viewBox="0 0 256 262"
              width="18px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M255.878,133.451 C255.878,122.717 255.007,114.884 253.122,106.761 L130.55,106.761 L130.55,155.209 L202.497,155.209 C201.047,167.249 193.214,185.381 175.807,197.565 L175.563,199.187 L214.318,229.21 L217.003,229.478 C241.662,206.704 255.878,173.196 255.878,133.451"
                  fill="#4285F4"
                />
                <path
                  d="M130.55,261.1 C165.798,261.1 195.389,249.495 217.003,229.478 L175.807,197.565 C164.783,205.253 149.987,210.62 130.55,210.62 C96.027,210.62 66.726,187.847 56.281,156.37 L54.75,156.5 L14.452,187.687 L13.925,189.152 C35.393,231.798 79.49,261.1 130.55,261.1"
                  fill="#34A853"
                />
                <path
                  d="M56.281,156.37 C53.525,148.247 51.93,139.543 51.93,130.55 C51.93,121.556 53.525,112.853 56.136,104.73 L56.063,103 L15.26,71.312 L13.925,71.947 C5.077,89.644 0,109.517 0,130.55 C0,151.583 5.077,171.455 13.925,189.152 L56.281,156.37"
                  fill="#FBBC05"
                />
                <path
                  d="M130.55,50.479 C155.064,50.479 171.6,61.068 181.029,69.917 L217.873,33.943 C195.245,12.91 165.798,0 130.55,0 C79.49,0 35.393,29.301 13.925,71.947 L56.136,104.73 C66.726,73.253 96.027,50.479 130.55,50.479"
                  fill="#EB4335"
                />
              </g>
            </svg>
          </span>
          <button
            class="login-provider-btn"
          >
            Continue With Google
          </button>
        </div>
        <div
          class="login-provider-btn-container"
        >
          <span
            class="provider-logo"
            id="apple-logo"
          >
            <svg
              height="18px"
              preserveAspectRatio="xMidYMid"
              version="1.1"
              viewBox="0 0 256 315"
              width="18px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M213.803394,167.030943 C214.2452,214.609646 255.542482,230.442639 256,230.644727 C255.650812,231.761357 249.401383,253.208293 234.24263,275.361446 C221.138555,294.513969 207.538253,313.596333 186.113759,313.991545 C165.062051,314.379442 158.292752,301.507828 134.22469,301.507828 C110.163898,301.507828 102.642899,313.596301 82.7151126,314.379442 C62.0350407,315.16201 46.2873831,293.668525 33.0744079,274.586162 C6.07529317,235.552544 -14.5576169,164.286328 13.147166,116.18047 C26.9103111,92.2909053 51.5060917,77.1630356 78.2026125,76.7751096 C98.5099145,76.3877456 117.677594,90.4371851 130.091705,90.4371851 C142.497945,90.4371851 165.790755,73.5415029 190.277627,76.0228474 C200.528668,76.4495055 229.303509,80.1636878 247.780625,107.209389 C246.291825,108.132333 213.44635,127.253405 213.803394,167.030988 M174.239142,50.1987033 C185.218331,36.9088319 192.607958,18.4081019 190.591988,0 C174.766312,0.636050225 155.629514,10.5457909 144.278109,23.8283506 C134.10507,35.5906758 125.195775,54.4170275 127.599657,72.4607932 C145.239231,73.8255433 163.259413,63.4970262 174.239142,50.1987249"
                  fill="#000000"
                />
              </g>
            </svg>
          </span>
          <button
            class="login-provider-btn"
          >
            Continue With Apple
          </button>
        </div>
        <input
          class="form-input "
          name="username"
          placeholder="Username"
          required=""
          type="text"
          value=""
        />
        <input
          class="form-input "
          name="password"
          placeholder="Password"
          required=""
          type="password"
          value=""
        />
        <button
          class="submit-btn "
          data-testid="login-btn"
        >
          Log In
        </button>
        <p
          class="smaller"
        >
          Forgot your username or password ?
        </p>
        <p
          class="smaller"
        >
          New to Reddit? 
          <a
            href="/signup"
          >
            Sign up
          </a>
        </p>
      </form>
    </main>
  </div>
</div>
`;

exports[`Login works renders 1`] = `
<div>
  <div
    class="login-page"
  >
    <aside
      class="artwork"
    />
    <main
      class="login-and-signup"
    >
      <h4>
        Log in
      </h4>
      <p
        class="user-agreement smaller"
      >
        By continuing, you agree to our User Agreement and Privacy Policy.
      </p>
      <form
        class="login-form"
      >
        <div
          class="login-provider-btn-container"
        >
          <span
            class="provider-logo"
            id="google-logo"
          >
            <svg
              height="18px"
              preserveAspectRatio="xMidYMid"
              version="1.1"
              viewBox="0 0 256 262"
              width="18px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M255.878,133.451 C255.878,122.717 255.007,114.884 253.122,106.761 L130.55,106.761 L130.55,155.209 L202.497,155.209 C201.047,167.249 193.214,185.381 175.807,197.565 L175.563,199.187 L214.318,229.21 L217.003,229.478 C241.662,206.704 255.878,173.196 255.878,133.451"
                  fill="#4285F4"
                />
                <path
                  d="M130.55,261.1 C165.798,261.1 195.389,249.495 217.003,229.478 L175.807,197.565 C164.783,205.253 149.987,210.62 130.55,210.62 C96.027,210.62 66.726,187.847 56.281,156.37 L54.75,156.5 L14.452,187.687 L13.925,189.152 C35.393,231.798 79.49,261.1 130.55,261.1"
                  fill="#34A853"
                />
                <path
                  d="M56.281,156.37 C53.525,148.247 51.93,139.543 51.93,130.55 C51.93,121.556 53.525,112.853 56.136,104.73 L56.063,103 L15.26,71.312 L13.925,71.947 C5.077,89.644 0,109.517 0,130.55 C0,151.583 5.077,171.455 13.925,189.152 L56.281,156.37"
                  fill="#FBBC05"
                />
                <path
                  d="M130.55,50.479 C155.064,50.479 171.6,61.068 181.029,69.917 L217.873,33.943 C195.245,12.91 165.798,0 130.55,0 C79.49,0 35.393,29.301 13.925,71.947 L56.136,104.73 C66.726,73.253 96.027,50.479 130.55,50.479"
                  fill="#EB4335"
                />
              </g>
            </svg>
          </span>
          <button
            class="login-provider-btn"
          >
            Continue With Google
          </button>
        </div>
        <div
          class="login-provider-btn-container"
        >
          <span
            class="provider-logo"
            id="apple-logo"
          >
            <svg
              height="18px"
              preserveAspectRatio="xMidYMid"
              version="1.1"
              viewBox="0 0 256 315"
              width="18px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M213.803394,167.030943 C214.2452,214.609646 255.542482,230.442639 256,230.644727 C255.650812,231.761357 249.401383,253.208293 234.24263,275.361446 C221.138555,294.513969 207.538253,313.596333 186.113759,313.991545 C165.062051,314.379442 158.292752,301.507828 134.22469,301.507828 C110.163898,301.507828 102.642899,313.596301 82.7151126,314.379442 C62.0350407,315.16201 46.2873831,293.668525 33.0744079,274.586162 C6.07529317,235.552544 -14.5576169,164.286328 13.147166,116.18047 C26.9103111,92.2909053 51.5060917,77.1630356 78.2026125,76.7751096 C98.5099145,76.3877456 117.677594,90.4371851 130.091705,90.4371851 C142.497945,90.4371851 165.790755,73.5415029 190.277627,76.0228474 C200.528668,76.4495055 229.303509,80.1636878 247.780625,107.209389 C246.291825,108.132333 213.44635,127.253405 213.803394,167.030988 M174.239142,50.1987033 C185.218331,36.9088319 192.607958,18.4081019 190.591988,0 C174.766312,0.636050225 155.629514,10.5457909 144.278109,23.8283506 C134.10507,35.5906758 125.195775,54.4170275 127.599657,72.4607932 C145.239231,73.8255433 163.259413,63.4970262 174.239142,50.1987249"
                  fill="#000000"
                />
              </g>
            </svg>
          </span>
          <button
            class="login-provider-btn"
          >
            Continue With Apple
          </button>
        </div>
        <input
          class="form-input "
          name="username"
          placeholder="Username"
          required=""
          type="text"
          value=""
        />
        <input
          class="form-input "
          name="password"
          placeholder="Password"
          required=""
          type="password"
          value=""
        />
        <button
          class="submit-btn "
          data-testid="login-btn"
        >
          Log In
        </button>
        <p
          class="smaller"
        >
          Forgot your username or password ?
        </p>
        <p
          class="smaller"
        >
          New to Reddit? 
          <a
            href="/signup"
          >
            Sign up
          </a>
        </p>
      </form>
    </main>
  </div>
</div>
`;
